RFC-0006-nan-definition.md

```diff
@@ -0,0 +1,179 @@
+https://www.ibm.com/docs/en/db2-for-zos/13?topic=numbers-double-precision-floating-point-double-float
+https://www.ibm.com/docs/en/db2-for-zos/13?topic=numbers-decimal-floating-point-decfloat
+https://www.ibm.com/docs/en/db2-for-zos/13?topic=comparison-numeric-comparisons 
+
+
+## Goals
+
+* Define a consistent behavior for ordering and comparisons of Nans in Presto
+* Make behavior easy to implement and debug
+* Make behavior easy for users to understand
+* Align behavior with Velox for consistency across Java and C++ workers
+
+## Proposed Implementation
+
+We propose defining ordering and comparisons for NaNs such that `NaN = NaN` and sorts biggest.
+
+With this proposal we stray from the IEEE definition and follow the example of postgresql and others.  In this case for all equality purposes (including =, distinctness, grouping, join keys, etc.) we would define NaN=NaN.  We would sort NaN as greater than infinity for all purposes, including for min/max and > and < comparisons.
+
+This was by far the most common behavior across the DBS we looked at, including Spark, which at Meta is the engine we care most about compatibility with.  It is not IEEE-754 compliant, but no other DB we looked at was (either they did not support NaNs or they had at least some features where they strayed from IEEE compliance). And given how erratically we were behaving previously, it is safe to say that our users have not been relying on IEEE-754 compliance (or if they have, they certainly have not been getting it). This definition is consistent across all equality,comparison, and ordering operations meaning that users can always know what to expect, and though it would require some changes in Presto, the implementation should be pretty straight forward and reduces the chances of bugs. Velox has a proposal for the same: https://github.com/facebookincubator/velox/pull/7237.
+
+The disadvantage is that we aren't strictly following the IEEE-754 definitions.
+
+We propose changing the behavior of all of our functinos and operators to match this definition (see the spreadsheet linked in the description of Presto's current behavior for the full list). Notably, this will involve changing basic operators (<, >, =, <>) and joins.  another important thing to call out is that Presto may read files and metadata that were written by other systems and it can write files and metadata that will be read by other systems. It is important to ensure that, for example, when Presto writes stats for max values it follows the expectations for the service it is writing too, and when it reads max values, it knows whether those will adhere to its own NaN ordering definition before using that metadata for its own optimization.
```

Comment:
```suggestion
We propose changing the behavior of all of our functinos and operators to match this definition (see the spreadsheet linked in the description of Presto's current behavior for the full list). Notably, this will involve changing basic operators (<, >, =, <>) and joins. Another important concern is that Presto may read files and metadata that were written by other systems, and it can write files and metadata that will be read by other systems. When Presto writes stats for max values it should follow the expectations for the service it is writing to, and when it reads max values it should know whether those will adhere to its own NaN ordering definition before using that metadata for its own optimization.
```